components/core/src/clp/ffi/KeyValuePairLogEvent.cpp

```diff
@@ -167,4 +404,73 @@ auto KeyValuePairLogEvent::create(
     }
     return KeyValuePairLogEvent{std::move(schema_tree), std::move(node_id_value_pairs), utc_offset};
 }
+
+auto KeyValuePairLogEvent::serialize_to_json(
+) const -> OUTCOME_V2_NAMESPACE::std_result<nlohmann::json> {
+    if (m_node_id_value_pairs.empty()) {
+        return nlohmann::json::object();
+    }
+
+    auto const sub_schema_tree_bitmap_ret{
+            get_sub_schema_tree_bitmap(m_node_id_value_pairs, *m_schema_tree)
+    };
+    if (sub_schema_tree_bitmap_ret.has_error()) {
+        return sub_schema_tree_bitmap_ret.error();
+    }
+    auto const& sub_schema_tree_bitmap{sub_schema_tree_bitmap_ret.value()};
+
+    auto json_root = nlohmann::json::object_t();
+
+    bool json_exception_captured{false};
+    auto json_exception_handler = [&]([[maybe_unused]] nlohmann::json::exception const& ex
+                                  ) -> void { json_exception_captured = true; };
+    using DfsIterator = SchemaTreeDfsIterator<decltype(json_exception_handler)>;
+
+    // Note: we explicitly use `std::stack` which has `std::deque` as the underlying container.
+    // Otherwise, we have to implement move semantics for `DfsIterator` to enable vector growth.
```

Comment:
```suggestion
    // NOTE: We use a `std::stack` (which uses `std::deque` as the underlying container) instead of
    // a `std::vector` to avoid implementing move semantics for `DfsIterator` (required when the
    // vector grows).
```